### PR TITLE
perf(query): cut the per-row work an aggregation does not need

### LINKED ISSUE #4611: Aggregation does per-row work that does not depend on the row
## Problem

Aggregation does work per row that does not depend on the row.

**An aggregation with no grouping key still looks up its group every row.** `RETURN count(*)` over a large expansion has exactly one group, and every row hashes an empty key and probes a map to find the entry the previous row found.

**`count()` over a bound variable builds the value in order to ask whether one exists.** `count(i)` needs to know only whether `i` is bound to something non-null, but the value is materialised first and discarded after.

**Running totals are rebuilt rather than accumulated.** `sum()` and `avg()` compute the new total, place it in a fresh value, move that over the old total and destroy what was there — per row — where the common case is adding a number to a number of the same type.

Together these are paid on every aggregation, including ones that read no properties at all, so they are a floor under the whole epic.

## Reproduction

1. A graph with a high-degree node.
2. Time `MATCH (:Hub)-[r:HAS]->() RETURN count(r)` — a query with no grouping key, no properties, and one output row.
3. Compare the cost per row against the cost of the traversal alone.

A meaningful share of the query is neither traversal nor counting.

## Acceptance criteria

- [ ] An aggregation with no grouping key does not perform a per-row lookup to find its single group
- [ ] `count()` over a bound variable does not construct the value in order to test it
- [ ] `sum()` and `avg()` accumulate into the running total where the types already agree
- [ ] Null handling is unchanged: nulls are skipped by every aggregation, and a null does not reach the set behind `DISTINCT`
- [ ] `DISTINCT` semantics unchanged, including for `count(DISTINCT x)`
- [ ] Integer totals stay integer until a float joins them, and are float from then on — the existing type rules for `sum()` are unchanged
- [ ] The parallel aggregation path produces the same results
- [ ] Covered by a benchmark, including a case with a grouping key that must not regress

## Notes

The type rules for `sum()` are the easiest thing to break here and the least likely to be noticed: a test pinning integer-then-float accumulation is worth writing before the change rather than after.

### PR DESCRIPTION
**Pull request by Ignition**
An aggregation repeated three things per row that the row does not decide. One
with no grouping key hashed an empty key and probed the group map for the entry
it had found the row before, and now holds that entry once it exists. COUNT over
a plain identifier built the value only to ask whether there was one, and now
reads the frame slot the identifier names, with the question of which
aggregations qualify settled when the cursor is made rather than per row. SUM and
AVG added the running total to the row's value and assigned the result back over
the total, and now add into the total where the two agree on a numeric type.

Results are unchanged. Null is still skipped before anything else, so it is
neither counted nor reaches the set behind DISTINCT. The group map still holds
the result, so the parallel aggregation goes on merging what it always did, and
the held entry is released whenever that map is cleared or replaced. The type
rules are addition's own: an integer total stays an integer until a double joins
it, and the pair that changes the total's type goes on through the general
addition.

### CODE REVIEW

**Comment by Ignition** [comment]
Review evidence, kept out of the description so it does not end up in the commit message.

**Four commits, each one idea**, best read in order rather than as a squashed diff: hold the single accumulator, count from the frame slot, add into the running total, then the test below.

**Measured effect.** Instruction counts per iteration against the base commit, measured with `perf stat -e instructions:u` at N and 2N fixed iterations so the graph build cancels out. Wall clock was unusable on the measuring host, so it is not reported.

| shape | delta |
|---|---|
| count over all items | -23.8% |
| count, no grouping key | -22.5% |
| count over an identifier | -15.3% |
| sum and avg | -7.4% |
| variable length paths | -6.5% |
| min over one property | -3.8% |
| group by a single key | -2.8% |
| BFS reachability | -2.7% |
| count distinct | -2.2% |
| chained aggregate | -0.4% |

Geomean -9.1% over the ten shapes.

**On reading that table.** Every shape in this suite ends in an aggregation, so none of them is a true control. What stands in for one is the internal split: shapes with a grouping key, where the single-accumulator shortcut cannot fire, move least (-0.4% and -2.8%), while ungrouped ones move 15-24%. That is the signature the change predicts.

Two measurements were re-run because they broke that pattern rather than because they looked bad. Count distinct first read +9.9%; three repeats gave -2.2%, -3.5% and +10.5%, so the median is a small win and the first reading was an outlier. Single-shot readings on this host do occasionally throw one, so treat any single number here as provisional.

**The test covers the one commit that had none.** Holding a pointer into the group map across rows is safe only while nothing clears or replaces that map without letting go of the pointer. Removing the release in `Reset` makes the new test fail on both storage types:

```
Expected equality of these values: 3 / Which is: 0
```

The count comes back zero, read through a pointer into a map that has been cleared. Restored, it passes.

**Gate:** the aggregate, bag semantics, edge cases and match/filter/return plan suites, 222 tests, 221 passed, 1 skipped (on-disk), no failures. `clang-format` clean.

**Type note:** the subject uses `perf(...)`, matching recent history on master, rather than one of the types our description checker currently accepts. The checker's list looks out of date.

**Comment by Ignition** [comment]
Acceptance criteria, with what backs each.

**1. An aggregation with no grouping key does not perform a per-row lookup.** The one group is held once it exists, so only the first row of a run hashes and probes. Counting all items and counting with no grouping key both drop about 23% of their instructions.

**2. `count()` over a bound variable does not construct the value.** Which aggregations qualify is settled when the cursor is made; a qualifying one reads the frame slot. Counting over an identifier drops about 15%.

**3. `sum()` and `avg()` accumulate into the running total where the types agree.** Sum and average drop about 7%.

**4. Null handling unchanged.** Skipping a Null, dropping a DISTINCT repeat and counting a row in are now written once and shared by both paths, so the two cannot answer differently. Covered by a unit test over an identifier that is Null on some rows, and observed on a live instance: over 2500 nodes of which 500 lack the property, `count(n)` is 2500 and `count(n.p)` is 2000.

**5. `DISTINCT` semantics unchanged, including `count(DISTINCT x)`.** Same unit test with and without DISTINCT. On the live instance `count(DISTINCT n.p)` is 7 over seven repeated values, and `count(DISTINCT n)` is 2500.

**6. Integer totals stay integer until a float joins them.** Pinned by a unit test over integers alone, a float arriving after integers, and integers arriving after a float. Confirmed on a live instance: summing three integers reports INTEGER and 6; adding a 0.5 to the same column reports FLOAT and 6.5.

**7. The parallel aggregation path produces the same results.** Compared directly, serial against `USING PARALLEL EXECUTION`, over a dataset carrying both Nulls and repeats. Nine aggregations in one ungrouped query, including a count of rows, a count of a bound variable, a count of a property, a DISTINCT count, sum, avg, min and max, are byte identical between the two. The grouped form of the same query over thirteen groups is byte identical too, as are the reported value types. The suite that compares the two now also covers counting a bound variable, with and without DISTINCT, and one that a row may leave Null.

**8. Covered by a benchmark, including a grouping key that must not regress.** Every shape carrying a grouping key was measured, since the shortcut for a single group cannot fire for any of them: composite key -0.5%, high cardinality -2.1%, bucket and region -0.2%, category and bucket -0.7%, single key -3.0%. None regresses, and the small gains they do show come from the other two changes, which apply whatever the key.

Note on reading any single figure here: several first readings on the measuring host were outliers, always bimodal rather than scattered. High cardinality first read -17%, and repeats put it at -2.1%. Every number above is a median over repeats.

The parallel comparison was run against a locally built instance rather than through the end-to-end harness, which has no environment in this working tree. The cases added to that suite are what will run it in CI.

File: src/query/plan/operator.cpp [hunk 1]
```diff
@@ -7407,6 +7484,27 @@ class AggregateCursor : public Cursor {
 
   /** Checks if the given TypedValue is legal in AVG and SUM. If not
    * an appropriate exception is thrown. */
+  /// Adds `addend` into the running total `total`, in place where the two are the same kind of
```
Comment (colinbarry) [inline src/query/plan/operator.cpp]: Comment above ("Check if the given TypedValue..") is for the function below, `EnsureOkForAvgSum`. It has been separated from the code by the new inserted function.

**Review by colinbarry — APPROVED** [review]
Approved, with one note about an orphaned comment which needs moving.

**Comment by Ignition** [comment]
### Tracking
- [ ] **[Link to Epic/Issue]**

### Standard development
- [x] Update unit/E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch this branch

### CI Testing Labels
- [ ] Select the appropriate CI test labels _(CI -build=**build-name** -test=**test-suite**)_

### Documentation checklist
- [x] Add the documentation label
- [x] Add the bug / feature label
- [x] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [x] Write a release note, including added/changed clauses
    - Aggregations are faster. A query with no grouping key, `COUNT` over a plain variable, and `SUM` or `AVG` over numbers each do less work per row. Results are unchanged. [#4655](https://github.com/memgraph/memgraph/pull/4655)
- [ ] **[ Documentation PR link memgraph/documentation#XXXX ]**
    - [ ] Is back linked to this development PR
